### ISSUE TITLE
Reorganise component files?

### ISSUE THREAD

**Issue by callumlocke** [Financial-Times/next-stream-page#868]
Q for @i-like-robots and @andygnewman

The app seems to have evolved three different ways of organising components...

Everything in ./components:

```
./components/stream-item/index.js
./components/stream-item/_main.scss
```

Everything in ./client/components:

```
./client/components/video-stream/index.html
./client/components/video-stream/index.js
./client/components/video-stream/main.scss
```

Split between the two:

```
./components/card-open-in-stream/index.js
./client/components/card-open-in-stream/_main.scss
```

Can we just go with the first strategy for everything, regardless of whether it's used isomorphically or not?

**Comment by i-like-robots** [comment]
Yes, I've brought this up as well. I don't mind the separation of client-only and server/client so much but combining the two is not so fun. I have a feeling there are a few components that have fallen out of use or into disrepair that could be cleaned up which could be done as part of this.

**Comment by callumlocke** [comment]
One issue I ran into: changes in the root components dir won't trigger watch driven builds of the client bundles... Tried using flags on the nbt build command, to make it watch the root components dir as well as client, but couldn't get it working, maybe I'm doing it wrong.

**Comment by matthew-andrews** [comment]
> Tried using flags on the nbt build command, to make it watch the root components dir as well as client, but couldn't get it working…
Come find me tomorrow… I can probably help bash the build tools into shape

**Comment by ironsidevsquincy** [comment]
Isn't the pattern shared (react) components live in `/components`, whereas strictly client-side 'components' live in `/client/components`? Means we can exclude `/client/components` from our nodemon watch

**Comment by callumlocke** [comment]
Ah I can see that's useful. I didn't realise there were any client-only components, thought everything got prerendered on the server. 

What about the thing of splitting up a component's files, like card-open-in-stream (SCSS file is in client; JS is not)? Think that's the only component that's split up like this currently

**Comment by ironsidevsquincy** [comment]
I think the confusion is because we kind of call anything that runs on the client a 'component', and we also call strict react modules 'components' too

I think it makes sense for a component's styling to live in the same dir, and in general we're following this rule

**Comment by callumlocke** [comment]
Can you give me an example of one that runs on the client but isn't a React component?

**Comment by ironsidevsquincy** [comment]
Everything in here we've called a 'component' - https://github.com/Financial-Times/next-front-page/tree/master/client/components

**Comment by callumlocke** [comment]
OK I can see it makes sense to keep those bits of code within the `client` dir.

Is there a way we can make the distinction clearer with the naming?

e.g. rename `./components` to `./shared-components`?

**Comment by ironsidevsquincy** [comment]
I think we should use `components` for react components (they will be shared by virtue of the dir being in the root of the project, and allows `/client/components` for any client-side-only react components), and move any non-react code to `/client/modules`, `/client/lib`, whatever

**Comment by callumlocke** [comment]
Sounds good to me. If no one has any objections I'll go with that

**Comment by i-like-robots** [comment]
For reference the newsroom tech team use:

```
client/
shared/
server/
```

**Comment by callumlocke** [comment]
I like that, intent is much clearer.

So we'd have `/client/components` and `/shared/components`?

And the clientside watcher would be configured to watch `./{client,shared}/**/*.scss` and `./{client,shared}/**/*.js`

And the serverside watcher would watch `./{server,shared}/**/*.js`, along with whatever else it watches

?

**Comment by ironsidevsquincy** [comment]
:+1:

**Comment by i-like-robots** [comment]
Sounds grand

**Comment by andygnewman** [comment]
Like it

**Comment by matthew-andrews** [comment]
can we add a note in the conventions part of the dev docs?
http://financial-times.github.io/next/docs/developer-guide/#conventions

Something like:-
> We try to group JavaScript, CSS and template files together as ‘components’.  Components should have their own folders named after the component and those folder should contain at least one of an `index.js` (inclusive) or a `_main.scss` file.  A component is server side only should go in a parent folder in the root of the project called `server/`, client side only inside `client/` and components that need to run in both places inside `shared/`.

**Comment by matthew-andrews** [comment]
Also publicise the PR to update the docs in the #ft-next-dev channel please :smile:

**Comment by callumlocke** [comment]
what did you mean by `(inclusive)`? Just realised I copied that over without knowing what it meant :)

**Comment by matthew-andrews** [comment]
‘exclusive or’ (strictly A or B) vs ‘inclusive or’ (A or B, or, A and B)

**Comment by matthew-andrews** [comment]
I think my inclusive in the wrong (/confusing) place…  probably should be A or B (inclusive)

**Comment by ironsidevsquincy** [comment]
We now have this structure on the front page

https://github.com/Financial-Times/next-front-page

However, we've gone with naming the sass and js files the same as the component, e.g

```
tabs/_tabs.scss
tabs/tabs.js
```

Makes searching for files (thousands of `_main.scss` vs one `_tabs.scss`), and knowing what file is being referenced in a web inspector a lot easier

**Comment by callumlocke** [comment]
closed by c63754787e5753c8c62fc03a598d44e89cd26c8d
